### ISSUE TITLE
object detection example: invalid object detection variable list

### ISSUE THREAD

**Issue by haghad** [sassoftware/python-dlpy#92]
I'm trying to follow the simple object detection example,
and I'm getting the following error in the screenshot below.
I'm not sure what went wrong, or how to examine the "object detection variable list" more closely to debug this problem.
Note that one difference in my code versus the example is that I'm not using pre-trained weights, but I don't imagine that's relevant.
Any advice is much appreciated.

![image](https://user-images.githubusercontent.com/44681869/50184901-5e55ef80-02e4-11e9-8c24-ac68f39d14ff.png)

**Comment by sasxianhu** [comment]
The cell 35 is not correct. The input data should like this:
_nobjects_, _object0_, _object0_x, _object0_y,_object0_width, _object0_height…


The first column means how many objects the image contains. If it is 5, then you have to have 5*5 columns to contain the object information. Each object is defined by class name (e.g., _object0_), bounding box information (e.g., _object0_x, object0_y,object0_width, object0_height)

**Comment by haghad** [comment]
I don't understand...
That cell is exactly the same as cell 10 in the example:
https://github.com/sassoftware/python-dlpy/blob/master/examples/object_detection/SimpleObjectDetection.ipynb

But nevertheless, are you saying the input_vars list should be the same as the targets list?
I tried changing my code so that inputVars = targets, but that leads to the same error.

Also, in my case I have one type of object and up to 3 of those objects can be found in a single image.

**Comment by sasxianhu** [comment]
Do a columInfo call on your table to check the columns first.
And  then print your targets.

Check whether they match or not.

**Comment by haghad** [comment]
You are correct that my targets list does not match the trainset columns...

![image](https://user-images.githubusercontent.com/44681869/50186375-94956e00-02e8-11e9-8026-1142140769ce.png)

I suppose I could subset this trainset table, to drop the other columns, but then how would the model get access to the actual training images?  Aren't the training images in this very table?
Thanks.

**Comment by sasxianhu** [comment]
I assume the maximum number of objects in the training data is 3. Then the targets variable list seems correct.

Please also check the column types. _nojbects_ must be numerical, *_x, *_y, *_width, *_height must be numerical as well.

**Comment by sasxianhu** [comment]
BTW: _object*_ must be char type for your object labels.

**Comment by bensloane** [comment]
Can you also post the cell used to construct the model?

**Comment by haghad** [comment]
Re: "BTW: _object*_ must be char type for your object labels."
"_nojbects_ must be numerical, *_x, *_y, *_width, *_height must be numerical as well"
It looks right, here's a screenshot:
![image](https://user-images.githubusercontent.com/44681869/50234654-3027ec80-0384-11e9-9350-430369e87aea.png)

**Comment by haghad** [comment]
> Can you also post the cell used to construct the model?

![image](https://user-images.githubusercontent.com/44681869/50234953-ce1bb700-0384-11e9-8574-1efa469ca47b.png)

**Comment by bensloane** [comment]
Ok so it looks like object2 coordinates are varchar... can you try changing to double?

**Comment by haghad** [comment]
Good eye, well-spotted!  
Could you please tell me how to change the type of those columns?
(Note that I created this CasTable from a csv file, and
after manually checking the csv file, I'm a bit puzzled about why 
for object2's box coordinates they got uploaded as a char instead of double.
I mean it got all the other ones correct for object0 and object1, automatically.)
Thanks.

**Comment by sasxianhu** [comment]
You can specify the column types when you call upload_frame like this:


def get_imagenet_labels_table(conn):


    temp_name = random_name('new_label_table', 6)





    filename = os.path.join('datasources', 'imagenet_labels.csv')


    project_path = os.path.dirname(os.path.abspath(__file__))


    full_filename = os.path.join(project_path, filename)





    labels = pd.read_csv(full_filename, skipinitialspace=True, index_col=False)


    conn.upload_frame(labels, casout=dict(name=temp_name, replace=True),


                      importoptions={'vars':[


                          {'name': 'label_id', 'type': 'int64'},


                          {'name': 'label', 'type': 'char', 'length': 200}]})





    return conn.CASTable(temp_name)



The reason that the column types are changed to varchar is most likely caused by your csv files. Not sure how you represent the missing values.

**Comment by bensloane** [comment]
See more [here](https://go.documentation.sas.com/?cdcId=pgmsascdc&cdcVersion=9.4_3.4&docsetId=caspg&docsetTarget=compg-importoptionsparam.htm&locale=en#PYTHON.cascommon.prm-importOptions-alt_csvopts) for specifying format type during import. You can pass in as a list of dicts where the var name and format would be separate keys and the values would be the var name and format you'd like to enforce respectively.

For changing the format after table is already loaded it looks like you can use the alterTable() action on the underlying CASTable to change the column format although I haven't tried this. See [here](https://go.documentation.sas.com/?cdcId=pgmsascdc&cdcVersion=9.4_3.4&docsetId=caspg&docsetTarget=cas-table-altertable.htm&locale=en#PYTHON.cas-table-altertable-columns-format)

You could also write some datastep, see [here](http://support.sas.com/kb/24/590.html), to do this which would look something like the following I believe. 

`data new;
   char_var = '12345678';
   numeric_var = input(char_var, 8.);
run;
`

**Comment by haghad** [comment]
> For changing the format after table is already loaded it looks like you can use the alterTable() action on the underlying CASTable to change the column format although I haven't tried this. See [here](https://go.documentation.sas.com/?cdcId=pgmsascdc&cdcVersion=9.4_3.4&docsetId=caspg&docsetTarget=cas-table-altertable.htm&locale=en#PYTHON.cas-table-altertable-columns-format)

Yes, I want to change the column type after the table is already loaded, so how does one do that?
alterTable() doesn't seem to have the option to change "Type".  I tried playing around with changing "Format" but that's not the same thing.

**Comment by sasxianhu** [comment]
You cannot change the type…

**Comment by bensloane** [comment]
Apologies I misunderstood the doc format != type :) ...  What I've done in the past is use the dlJoin action but dlpy has the `create_object_detection_table`  which is really convenient. In your case since you have a csv of annotations I would use XQ's suggestion above and enforce the type when uploading. Can you try reloading with the suggestions above?

**Comment by haghad** [comment]
this problem is pretty much resolved by getting the CASTable from a csv file in different ways.
In investigating this problem, I've found three ways.
METHOD 1: CSV -> Pandas DF -> CASTable
using "conn.upload_frame()" method on the pandas data frame created a CAS Table with the correct data types
(interestingly, the code says that this method converts back to csv before uploading into CASTable)

METHOD 2: CSV -> CASTable
conn.read_csv() method

METHOD 3 (the method I originally used):
use conn.table.loadtable() however, as @bensloane discovered and pointed out to me, the "guessrows" parameter sometimes needs to be modified so that during import the correct datatype is inferred, which in the case of my csv, this was the issue.

I am no longer getting the original error...
however I am getting this error:
![image](https://user-images.githubusercontent.com/44681869/50250771-d3422b80-03af-11e9-94aa-795b76d2bcb0.png)
which I am certain has nothing to do with the original issue stated in this thread.
=)
Thanks for your help!

**Comment by sasxianhu** [comment]
Do not use guessRows since the types could be still wrong. Do method 1 always if you know your column types.
Your new error means your GPU machine has the old GPU driver version.

**Comment by sasxianhu** [comment]
close this now.
